### PR TITLE
Applications should match if no postcode provided

### PR DESCRIPTION
**Pull request by avinhurry**
## Context

**Note: Support has requested for this to be merged on or after 03/05/22 and should be merged after the job has run at 1pm**

It came to our attention that international candidates were not being flagged by the duplicate matching algorithm as they don't always enter a postcode. 

**Proposed plan of action:**

- Manually flag the new international matches using the new SQL script after 1pm (as thats when the service runs)
- Don't send the email or block submission on any of these by using the new switches, e.g. `UpdateDuplicateMatches.new(matches: GetDuplicateMatches.call, send_email: false, block_submission: false)`
- Support to manually action each match, blocking submission and sending the email as appropriate. 
- Going forwards international matches with have the email automated and submission blocked automatically, in the same way as the non international candidates.

## Changes proposed in this pull request

- Change the duplicate matching SQL query so that if there are no postcodes provided it still matches
- Add a switch so that we can manually flag all the duplicate matches without sending the email
- Add a switch so that we can manually flag all the duplicate matches without blocking submission

## Guidance to review

- Any flaws in the updates SQL script?

## Link to Trello card

https://trello.com/c/qHjKH8Nm/4634-consider-country-if-no-postcode-for-duplicate-matching

### CODE REVIEW

File: app/queries/get_duplicate_matches.rb [hunk 1]
```diff
@@ -10,21 +10,21 @@ def self.call
           submitted_at
         FROM application_forms application_details
         JOIN(
-          SELECT TRIM(UPPER(application_forms.last_name)) last_name, application_forms.date_of_birth, REPLACE(UPPER(application_forms.postcode), ' ', '') postcode
+          SELECT TRIM(UPPER(application_forms.last_name)) last_name, application_forms.date_of_birth, COALESCE(REPLACE(UPPER(application_forms.postcode), ' ', ''), '') postcode
           FROM application_forms
           WHERE application_forms.previous_application_form_id IS NULL
           GROUP BY TRIM(UPPER(application_forms.last_name)), application_forms.date_of_birth, REPLACE(UPPER(application_forms.postcode), ' ', '')
```
Comment (stevehook) [inline app/queries/get_duplicate_matches.rb:16]: When I tested it with nulls it worked perfectly but didn't find matches that differed only by whitespace, e.g. (' ' and NULL). I think you need the `COALESCE` in the `GROUP BY` here and in the main `SELECT DISTINCT` clause above too.

**Review by stevehook — COMMENTED** [review]
Looks really good. When testing I think I found an issue with an edge case (comments inline).

File: app/queries/get_duplicate_matches.rb (hunk 1 above)
Comment (avinhurry) [reply app/queries/get_duplicate_matches.rb]: good spot. We've fixed and added a test for it.

File: app/queries/get_duplicate_matches.rb [hunk 2]
```diff
@@ -10,21 +10,21 @@ def self.call
...
           HAVING (count(*) > 1)
         ) duplicate_attributes
-        ON REPLACE(UPPER(application_details.postcode), ' ', '') = duplicate_attributes.postcode
+        ON COALESCE(REPLACE(UPPER(application_details.postcode), ' ', ''), '') = duplicate_attributes.postcode
```
[the remark this answers is not in the document]
Comment (avinhurry) [reply app/queries/get_duplicate_matches.rb:19]: Yes if there is no postcode, the match will be based on last name and DOB.

**Review by stevehook — APPROVED** [review]
:+1:

**Comment by avinhurry** [comment]
Thanks @stevehook. I think we can merge this after 1pm today, then run the job manually in production, setting `block_submission` and `send_email` to false. Shall we pair on this @tomas-stefano?
